<a href="https://colab.research.google.com/github/mehrnaznz/MPM_200_MehrnazNiazi/blob/main/last_session_homework-Mehrnaz%20Niazi.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
df = pd.read_excel("/content/sample_data/Flavivirus host associations.xlsx")

In [ ]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 570 entries, 0 to 569
Data columns (total 10 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   sr no.                  570 non-null    int64  
 1   Virus                   570 non-null    object 
 2   Scientific Name         570 non-null    object 
 3   Authors                 570 non-null    object 
 4   Year                    430 non-null    object 
 5   Title/accession number  570 non-null    object 
 6   Journal                 425 non-null    object 
 7   Issue                   415 non-null    float64
 8   Volume                  404 non-null    object 
 9   Pages                   404 non-null    object 
dtypes: float64(1), int64(1), object(8)
memory usage: 44.7+ KB
None


In [ ]:
df.head()

sr no.         Virus     Scientific Name  \
0       3    Apoi virus       Rattus rattus   
1       2    Apoi virus    Myodes rufocanus   
2       0    Apoi virus  Apodemus argenteus   
3       1    Apoi virus  Apodemus speciosus   
4       4  Bagaza virus      Alectoris rufa   

                                             Authors  Year  \
0  C. M. Baak-Baak, N. Cigarroa-Toledo, O. M. Tor...  2016   
1                                        CDC Arbocat   NaN   
2                 J. S. Mackenzie and D. T. Williams  2009   
3                 J. S. Mackenzie and D. T. Williams  2009   
4  D. Buitrago, A. Rocha, C. Tena-Tomas, M. Vigo,...  2012   

                              Title/accession number  \
0  New Records of Ectoparasites Echinolaelaps ech...   
1                                        CDC Arbocat   
2  The zoonotic flaviviruses of southern, south-e...   
3  The zoonotic flaviviruses of southern, south-e...   
4  Real-time fluorogenic reverse transcription po...   

                     Journal  Issue Volume   Pages  
0  Southwestern Entomologist    NaN    NaN     NaN  
1                        NaN    NaN    NaN     NaN  
2     Zoonoses Public Health   56.0    6-7  338-56  
3     Zoonoses Public Health   56.0    6-7  338-56  
4         J Vet Diagn Invest   24.0      5  959-63

In [ ]:
df.columns.tolist()

['sr no.',
 'Virus',
 'Scientific Name',
 'Authors',
 'Year',
 'Title/accession number',
 'Journal',
 'Issue',
 'Volume',
 'Pages']

In [ ]:
df.dtypes

sr no.                      int64
Virus                      object
Scientific Name            object
Authors                    object
Year                       object
Title/accession number     object
Journal                    object
Issue                     float64
Volume                     object
Pages                      object
dtype: object

# Count total number of entries in the dataset

In [ ]:
total_entries = len(df)
print(f"Total number of entries in the dataset: {total_entries}")

Total number of entries in the dataset: 570


# Count unique viruses, List of unique viruses, Count of each unique virus

In [ ]:
unique_viruses_count = df['Virus'].nunique()
print(f"Total number of unique viruses: {unique_viruses_count}")

unique_virus_names = df['Virus'].unique()
print("\nList of unique virus names:")
for virus_name in unique_virus_names:
    print(f"- {virus_name}")

virus_counts = df['Virus'].value_counts()
print("\nOccurrences of each unique virus:")
print(virus_counts)

Total number of unique viruses: 30

List of unique virus names:
- Apoi virus
- Bagaza virus
- Banzi virus
- Bouboui virus
- Bussuquara virus
- Cacipacore virus
- Dakar bat virus
- Dengue virus
- Entebbe bat virus
- Iguape virus
- Ilheus virus
- Japanese encephalitis virus
- Koutango virus
- Kunjin virus
- Kyasanur forest virus
- Louping-ill virus
- Modoc virus
- Murray Valley encephalitis virus
- Omsk hemorrhagic fever virus
- Rio Bravo virus
- Rocio virus
- St. Louis Encephalitis virus
- Tembusu virus
- Tickborne encephalitis virus
- Uganda S virus
- Usutu virus
- Wesselsbron virus
- West Nile virus
- Yellow fever virus
- Zika virus

Occurrences of each unique virus:
Virus
West Nile virus                     194
Tickborne encephalitis virus         84
St. Louis Encephalitis virus         72
Usutu virus                          65
Dengue virus                         27
Japanese encephalitis virus          25
Louping-ill virus                    13
Yellow fever virus                   

# Count unique host species, List of unique host species, Count of host species for each virus

In [ ]:
unique_host_species_count = df['Scientific Name'].nunique()
print(f"Total number of unique host species: {unique_host_species_count}")

unique_host_species_names = df['Scientific Name'].unique()
print("\nList of unique host species names:")
for species_name in unique_host_species_names:
    print(f"- {species_name}")

Total number of unique host species: 419

List of unique host species names:
- Rattus rattus
- Myodes rufocanus
- Apodemus argenteus
- Apodemus speciosus
- Alectoris rufa
- Phasianus colchicus
- Mastomys natalensis
- Mesocricetus auratus
- Papio papio
- Equus caballus
- Alouatta belzebul
- Proechimys guyannensis
- Chaerephon pumilus
- Taphozous perforatus
- Mops condylurus
- Scotophilus nigrita
- Otomops sp
- Carollia brevicauda
- Molossus molossus
- Myotis nigricans
- Natalus stramineus
- Rousettus leschenaultii
- Artibeus planirostris
- Caluromys philander
- Carollia perspicillata
- Didelphis marsupialis
- Hylaeamys megacephalus
- Marmosa demerarae
- Marmosa murina
- Marmosops parvidens
- Mesomys hispidus
- Oecomys sp
- Philander opossum
- Proechimys cuvieri
- Zygodontomys brevicauda
- Artibeus jamaicensis
- Artibeus lituratus
- Glossophaga soricina
- Artibeus sp
- Uroderma sp
- Macaca nemestrina
- Mus musculus
- Eidolon helvum
- Pipistrellus pipistrellus
- Rhinolophus sp
- Rousettus

# Most diverse viruses (viruses with most unique host species)

In [ ]:
virus_host_diversity = df.groupby('Virus')['Scientific Name'].nunique().sort_values(ascending=False)
print("Diversity of host species for each virus (sorted by most unique hosts):\n")
print(virus_host_diversity)

most_diverse_virus = virus_host_diversity.index[0]
num_unique_hosts = virus_host_diversity.iloc[0]
print(f"\nVirus associated with the most unique host species: {most_diverse_virus} ({num_unique_hosts} unique hosts)")

Diversity of host species for each virus (sorted by most unique hosts):

Virus
West Nile virus                     194
Tickborne encephalitis virus         83
St. Louis Encephalitis virus         72
Usutu virus                          65
Dengue virus                         27
Japanese encephalitis virus          25
Louping-ill virus                    13
Yellow fever virus                   11
Rio Bravo virus                       9
Zika virus                            7
Entebbe bat virus                     6
Kyasanur forest virus                 6
Ilheus virus                          5
Dakar bat virus                       5
Apoi virus                            4
Uganda S virus                        4
Modoc virus                           3
Murray Valley encephalitis virus      3
Wesselsbron virus                     3
Omsk hemorrhagic fever virus          3
Kunjin virus                          3
Bussuquara virus                      3
Bagaza virus                          2
B

In [ ]:
unique_host_species_count = df['Scientific Name'].nunique()
print(f"Total number of unique host species: {unique_host_species_count}")

unique_host_species_names = df['Scientific Name'].unique()
print("\nList of unique host species names:")
for species_name in unique_host_species_names:
    print(f"- {species_name}")

Total number of unique host species: 419

List of unique host species names:
- Rattus rattus
- Myodes rufocanus
- Apodemus argenteus
- Apodemus speciosus
- Alectoris rufa
- Phasianus colchicus
- Mastomys natalensis
- Mesocricetus auratus
- Papio papio
- Equus caballus
- Alouatta belzebul
- Proechimys guyannensis
- Chaerephon pumilus
- Taphozous perforatus
- Mops condylurus
- Scotophilus nigrita
- Otomops sp
- Carollia brevicauda
- Molossus molossus
- Myotis nigricans
- Natalus stramineus
- Rousettus leschenaultii
- Artibeus planirostris
- Caluromys philander
- Carollia perspicillata
- Didelphis marsupialis
- Hylaeamys megacephalus
- Marmosa demerarae
- Marmosa murina
- Marmosops parvidens
- Mesomys hispidus
- Oecomys sp
- Philander opossum
- Proechimys cuvieri
- Zygodontomys brevicauda
- Artibeus jamaicensis
- Artibeus lituratus
- Glossophaga soricina
- Artibeus sp
- Uroderma sp
- Macaca nemestrina
- Mus musculus
- Eidolon helvum
- Pipistrellus pipistrellus
- Rhinolophus sp
- Rousettus

1. How many entries are there in the dataset?
There are 570 entries in the dataset.

2. How many viruses are present in the dataset?
There are 30 unique viruses present in the dataset.

3. List those viruses.
The unique viruses are: Apoi virus, Bagaza virus, Banzi virus, Bouboui virus, Bussuquara virus, Cacipacore virus, Dakar bat virus, Dengue virus, Entebbe bat virus, Iguape virus, Ilheus virus, Japanese encephalitis virus, Koutango virus, Kunjin virus, Kyasanur forest virus, Louping-ill virus, Modoc virus, Murray Valley encephalitis virus, Omsk hemorrhagic fever virus, Rio Bravo virus, Rocio virus, St. Louis Encephalitis virus, Tembusu virus, Tickborne encephalitis virus, Uganda S virus, Usutu virus, Wesselsbron virus, West Nile virus, Yellow fever virus, and Zika virus.

4. How many rows of data are present for each virus?
West Nile virus: 194
Tickborne encephalitis virus: 84
St. Louis Encephalitis virus: 72
Usutu virus: 65
Dengue virus: 27
Japanese encephalitis virus: 25
Louping-ill virus: 13
Yellow fever virus: 11
Rio Bravo virus: 9
Zika virus: 7
Uganda S virus: 7
Entebbe bat virus: 6
Kyasanur forest virus: 6
Dakar bat virus: 5
Ilheus virus: 5
Apoi virus: 4
Modoc virus: 3
Murray Valley encephalitis virus: 3
Wesselsbron virus: 3
Omsk hemorrhagic fever virus: 3
Kunjin virus: 3
Bussuquara virus: 3
Bagaza virus: 2
Banzi virus: 2
Rocio virus: 2
Tembusu virus: 2
Koutango virus: 1
Iguape virus: 1
Bouboui virus: 1
Cacipacore virus: 1

5. How many host species are present in the data?
There are 419 unique host species present in the data.

6. Make a list of host species present.
**Answer (List of unique host species names):**
- Rattus rattus
- Myodes rufocanus
- Apodemus argenteus
- Apodemus speciosus
- Alectoris rufa
- Phasianus colchicus
- Mastomys natalensis
- Mesocricetus auratus
- Papio papio
- Equus caballus
- Alouatta belzebul
- Proechimys guyannensis
- Chaerephon pumilus
- Taphozous perforatus
- Mops condylurus
- Scotophilus nigrita
- Otomops sp
- Carollia brevicauda
- Molossus molossus
- Myotis nigricans
- Natalus stramineus
- Rousettus leschenaultii
- Artibeus planirostris
- Caluromys philander
- Carollia perspicillata
- Didelphis marsupialis
- Hylaeamys megacephalus
- Marmosa demerarae
- Marmosa murina
- Marmosops parvidens
- Mesomys hispidus
- Oecomys sp
- Philander opossum
- Proechimys cuvieri
- Zygodontomys brevicauda
- Artibeus jamaicensis
- Artibeus lituratus
- Glossophaga soricina
- Artibeus sp
- Uroderma sp
- Macaca nemestrina
- Mus musculus
- Eidolon helvum
- Pipistrellus pipistrellus
- Rhinolophus sp
- Rousettus spinalatus
- Macaca mulatta
- Equus ferus
- Molothrus bonariensis
- Sporophila caerulescens
- Hipposideros armiger
- Rhinolophus pusillus
- Vespertilio sinensis
- Bos taurus
- Bos gaurus
- Crocidura macowi
- Turdus merula
- Turdus iliacus
- Sturnus vulgaris
- Turdus philomelos
- Columba livia
- Sus scrofa
- Alouatta pigra
- Sus domesticus
- Myotis pilosus
- Scotophilus kuhlii
- Rhinolophus affinis
- Miniopterus schreibersii
- Ovis aries
- Murina aurata
- Padda oryzivora
- Spilopelia chinensis
- Gallus gallus
- Oriolus flavocinctus
- Pteropus sp
- Vandeleuria oleracea
- Macaca radiata
- Madromys blanfordi
- Semnopithecus entellus
- Capra aegagrus
- Capra hircus
- Apodemus sylvaticus
- Canis lupus
- Capreolus capreolus
- Cervus elaphus
- Lagopus lagopus
- Lepus timidus
- Sorex araneus
- Peromyscus maniculatus
- Egretta novaehollandiae
- Microtus gregalis
- Ondatra zibethicus
- Microtus oeconomus
- Molossus rufus
- Tadarida brasiliensis
- Eptesicus fuscus
- Desmodus rotundus
- Sturnira lilium
- Zonotrichia capensis
- Agelaioides badius
- Agelaius phoeniceus
- Alouatta sp
- Ateles paniscus
- Bradypus variegatus
- Bubo virginianus
- Buteogallus coronatus
- Butorides striata
- Callipepla gambelii
- Calomys musculinus
- Campylorhynchus brunneicapillus
- Caracara cheriway
- Carpodacus mexicanus
- Charadrius vociferus
- Choloepus hoffmanni
- Colaptes melanochloros
- Columbina passerina
- Columbina picui
- Conopophaga aurita
- Dendroica coronata
- Egretta thula
- Elaenia albiceps
- Falco sparverius
- Furnarius rufus
- Geothlypis trichas
- Geotrygon montana
- Hypocnemis cantator
- Icterus bullockii
- Ixobrychus exilis
- Lanius ludovicianus
- Lepus californicus
- Lophospingus pusillus
- Melospiza melodia
- Melozone aberti
- Mimus polyglottos
- Molothrus ater
- Myiobius barbatus
- Myrmotherula hauxwelli
- Nycticorax nycticorax
- Oligoryzomys nigripes
- Orcinus orca
- Passer domesticus
- Phalcoboenus chimango
- Philydor erythrocercum
- Pitangus sulphuratus
- Poospiza nigrorufa
- Porzana carolina
- Procyon lotor
- Pseudoseisura lophotes
- Pyrocephalus rubinus
- Saltator maximus
- Sapajus apella
- Sayornis nigricans
- Stelgidopteryx serripennis
- Toxostoma redivivum
- Turdus amaurochalinus
- Vermivora ruficapilla
- Zenaida asiatica
- Zenaida auriculata
- Zenaida macroura
- Tadorna sp
- Anas platyrhynchos
- Apodemus agrarius
- Myodes rutilus
- Alticola argentatus
- Apodemus pallipes
- Microtus ilaeus
- Bison bonasus
- Macaca sylvanus
- Apodemus flavicollis
- Microtus agrestis
- Microtus arvalis
- Apodemus sp
- Myodes glareolus
- Lagomorpha sp
- Sciuridae sp
- Acrocephalus dumetorum
- Anas crecca
- Bombycilla garrulus
- Carduelis carduelis
- Carduelis chloris
- Carduelis spinus
- Carpodacus erythrinus
- Dendrocopos major
- Emberiza citrinella
- Ficedula hypoleuca
- Luscinia calliope
- Luscinia luscinia
- Mergus merganser
- Parus ater
- Parus major
- Parus montanus
- Passer montanus
- Phoenicurus phoenicurus
- Phylloscopus collybita
- Phylloscopus proregulus
- Phylloscopus trochilus
- Pyrrhula pyrrhula
- Riparia riparia
- Sitta europaea
- Spatula clypeata
- Sylvia curruca
- Vanellus vanellus
- Ovis orientalis
- Accipiter gentilis
- Anthus trivialis
- Bonasa bonasia
- Carduelis flammea
- Clangula hyemalis
- Corvus corone
- Corvus monedula
- Crex crex
- Emberiza fucata
- Fringilla coelebs
- Fringilla montifringilla
- Fulica atra
- Jynx torquilla
- Lanius collurio
- Lanius cristatus
- Loxia curvirostra
- Melanitta fusca
- Motacilla alba
- Motacilla flava
- Pica pica
- Scolopax rusticola
- Spatula querquedula
- Troglodytes troglodytes
- Turdus pallidus
- Turdus pilaris
- Turdus ruficollis
- Arvicanthis niloticus
- Cricetomys gambianus
- Saxicola rubetra
- Andropadus virens
- Bycanistes sharpii
- Praomys sp
- Turdus libonyanus
- Aegolius funereus
- Erithacus rubecula
- Glaucidium passerinum
- Surnia ulula
- Hirundo rustica
- Parus caeruleus
- Strix nebulosa
- Ardea cinerea
- Clanga clanga
- Larus michahellis
- Merops apiaster
- Picus viridis
- Streptopelia decaocto
- Garrulus glandarius
- Ciconia ciconia
- Pandion haliaetus
- Equus asinus
- Acrocephalus scirpaceus
- Bubo bubo
- Circus aeruginosus
- Delichon urbicum
- Emberiza schoeniclus
- Falco tinnunculus
- Gypaetus barbatus
- Pavo cristatus
- Phoenicurus ochruros
- Strix aluco
- Strix uralensis
- Sylvia atricapilla
- Sylvia borin
- Sylvia communis
- Tyto alba
- Alcedo atthis
- Larosterna inca
- Serinus canaria
- Turdus sp
- Asio otus
- Caprimulgus europaeus
- Larus ridibundus
- Desmodillus auricularis
- Acrocephalus schoenobaenus
- Anthus pratensis
- Meleagris gallopavo
- Prunella modularis
- Dendroica petechia
- Icterus leucopteryx
- Leptotila jamaicensis
- Mniotilta varia
- Myiopagis cotta
- Parkesia noveboracensis
- Tiaris bicolor
- Turdus aurantius
- Turdus jamaicensis
- Tyrannus caudifasciatus
- Vireo modestus
- Anorrhinus tickelli
- Larus leucophthalmus
- Anser anser domesticus
- Platycercus sp
- Branta canadensis
- Mephitis mephitis
- Bonasa umbellus
- Ciconia sp
- Canis familiaris
- Sciurus carolinensis
- Cyanocitta cristata
- Corvus brachyrhynchos
- Gallus domesticus
- Bubulcus ibis
- Aegypius monachus
- Corvus frugilegus
- Coracornis raveni
- Cardinalis cardinalis
- Fulica americana
- Phalacrocorax sp
- Dumetella carolinensis
- Falco columbarius
- Falco punctatus
- Accipiter nisus
- Aquila adalberti
- Quiscalus quiscula
- Rodent sp
- Aquila fasciata
- Pelophylax ridibundus
- Aquila chrysaetos
- Sciurus niger
- Athene noctua
- Chaerephon bregullae
- Anaxyrus fowleri
- Parus atricapillus
- Arabitragus jayakari
- Corvus corax
- Crocodylus moreletii
- Panurus biarmicus
- Haliaeetus albicilla
- Hylocichla mustelina
- Toxostoma rufum
- Parus carolinensis
- Loriini sp
- Capra falconeri
- Pelecanus conspicillatus
- Quiscalus major
- Gymnogyps californianus
- Asio flammeus
- Coragyps atratus
- Buteo swainsoni
- Pelecanus occidentalis
- Cygnus buccinator
- Sternula albifrons
- Ictinia mississippiensis
- Phalacrocorax auritus
- Buteo regalis
- Calidris alba
- Sturnidae
- Charadrius melodus
- Haliaeetus leucocephalus
- Akodon mimus
- Accipitridae sp
- Larus smithsonianus
- Actitis macularius
- Sternula antillarum
- Pelecanus erythrorhynchos
- Phoenicopterus ruber
- Anatidae sp
- Nestor notabilis
- Camelus dromedarius
- Larus delawarensis
- Colaptes auratus
- Baeolophus bicolor
- Corvus ossifragus
- Accipiter striatus
- Carduelis tristis
- Turdus migratorius
- Anisognathus somptuosus
- Campylorhamphus trochilirostris
- Coereba flaveola
- Elaenia flavogaster
- Thraupis palmarum
- Turdus leucomelas
- Odocoileus virginianus
- Oreamnos americanus
- Rangifer tarandus
- Ursus maritimus
- Struthio camelus
- Accipiter cooperii
- Agelaius tricolor
- Aphelocoma californica
- Ardea alba
- Ardea herodias
- Buteo jamaicensis
- Buteo lineatus
- Callipepla californica
- Cistothorus palustris
- Dryobates nuttallii
- Euphagus cyanocephalus
- Falco peregrinus
- Geococcyx californianus
- Lonchura punctulata
- Melospiza lincolnii
- Melozone crissalis
- Myiarchus cinerascens
- Pheucticus melanocephalus
- Pica nuttalli
- Tachycineta bicolor
- Thryomanes bewickii
- Tyrannus verticalis
- Zonotrichia atricapilla
- Zonotrichia leucophrys
- Acrocephalus palustris
- Locustella luscinioides
- Saguinus geoffroyi
- Primate sp
- Erythrocebus patas
- Alouatta caraya
- Alouatta guariba
- Chlorocebus aethiops
- Alouatta sara
- Colobus guereza
- Aotus sp
- Callithrix jacchus
- Macaca fascicularis
- Pongo pygmaeus
- Sapajus libidinosus

7. Count the number of unique host species for each virus.
West Nile virus: 194 unique hosts
Tickborne encephalitis virus: 83 unique hosts
St. Louis Encephalitis virus: 72 unique hosts
Usutu virus: 65 unique hosts
Dengue virus: 27 unique hosts
Japanese encephalitis virus: 25 unique hosts
Louping-ill virus: 13 unique hosts
Yellow fever virus: 11 unique hosts
Rio Bravo virus: 9 unique hosts
Zika virus: 7 unique hosts
Entebbe bat virus: 6 unique hosts
Kyasanur forest virus: 6 unique hosts
Ilheus virus: 5 unique hosts
Dakar bat virus: 5 unique hosts
Apoi virus: 4 unique hosts
Uganda S virus: 4 unique hosts
Modoc virus: 3 unique hosts
Murray Valley encephalitis virus: 3 unique hosts
Wesselsbron virus: 3 unique hosts
Omsk hemorrhagic fever virus: 3 unique hosts
Kunjin virus: 3 unique hosts
Bussuquara virus: 3 unique hosts
Bagaza virus: 2 unique hosts
Banzi virus: 2 unique hosts
Rocio virus: 2 unique hosts
Tembusu virus: 2 unique hosts
Koutango virus: 1 unique host
Iguape virus: 1 unique host
Bouboui virus: 1 unique host
Cacipacore virus: 1 unique host

8. Which virus has the most number of hosts?
West Nile virus is associated with the most unique host species, with 194 unique hosts.

9. Which host has most number of viruses?
Equus caballus (horse)

In [ ]:
host_virus_diversity = df.groupby('Scientific Name')['Virus'].nunique().sort_values(ascending=False)
print("Diversity of viruses for each host species (sorted by most unique viruses):\n")
print(host_virus_diversity.head(10))

most_diverse_host = host_virus_diversity.index[0]
num_unique_viruses = host_virus_diversity.iloc[0]
print(f"\nHost species associated with the most unique viruses: {most_diverse_host} ({num_unique_viruses} unique viruses)")

Diversity of viruses for each host species (sorted by most unique viruses):

Scientific Name
Equus caballus          13
Rattus rattus            5
Mus musculus             5
Equus ferus              5
Sus scrofa               4
Gallus gallus            4
Ovis aries               4
Artibeus jamaicensis     4
Bos gaurus               4
Columba livia            4
Name: Virus, dtype: int64

Host species associated with the most unique viruses: Equus caballus (13 unique viruses)


#second data

In [ ]:
loc = "/content/sample_data/Handbook of the Birds of the World and BirdLife International Digital Checklist of the Birds of the World_Version_9.xlsx"
birds_meta = pd.read_excel(loc)
birds_meta.head()

Seq.  Subsp. Seq.             Order    Family name     Family Subfamily  \
0   1.0          0.0  STRUTHIONIFORMES  Struthionidae  Ostriches       NaN   
1   1.0          1.0               NaN            NaN        NaN       NaN   
2   1.0          2.0               NaN            NaN        NaN       NaN   
3   1.0          3.0               NaN            NaN        NaN       NaN   
4   1.0          4.0               NaN            NaN        NaN       NaN   

  Tribe     Common name             Scientific name       Authority  \
0   NaN  Common Ostrich            Struthio camelus  Linnaeus, 1758   
1   NaN             NaN   Struthio camelus syriacus             NaN   
2   NaN             NaN    Struthio camelus camelus             NaN   
3   NaN             NaN  Struthio camelus massaicus             NaN   
4   NaN             NaN  Struthio camelus australis             NaN   

  2024 IUCN Red List category Synonyms Alternative common names  \
0                          LC      NaN                      NaN   
1                         NaN      NaN                      NaN   
2                         NaN      NaN                      NaN   
3                         NaN      NaN                      NaN   
4                         NaN      NaN                      NaN   

                                 Taxonomic source(s)    SISRecID  \
0  del Hoyo, J., Collar, N.J., Christie, D.A., El...  45020636.0   
1                                                NaN         NaN   
2                                                NaN         NaN   
3                                                NaN         NaN   
4                                                NaN         NaN   

   SpcRecID\n(used before SISRecID)    SubsppID  
0                         1016860.0         NaN  
1                               NaN  45020636_1  
2                               NaN  45020636_2  
3                               NaN  45020636_3  
4                               NaN  45020636_4

#third data

In [ ]:
loc = "/content/sample_data/IUCN Mammals, Birds, Reptiles, and Amphibians.csv"
mammals_meta = pd.read_csv(loc, encoding='latin-1')
mammals_meta.head()

Species ID   Kingdom    Phylum     Class             Order       Family  \
0     42641.0  ANIMALIA  CHORDATA  MAMMALIA          RODENTIA      MURIDAE   
1  22687170.0  ANIMALIA  CHORDATA      AVES  CAPRIMULGIFORMES  TROCHILIDAE   
2     17879.0  ANIMALIA  CHORDATA  MAMMALIA          RODENTIA      MURIDAE   
3  47760825.0  ANIMALIA  CHORDATA  MAMMALIA          RODENTIA   CRICETIDAE   
4     42656.0  ANIMALIA  CHORDATA  MAMMALIA          RODENTIA  ABROCOMIDAE   

          Genus    Species                  Authority  Infraspecific rank  \
0     Abditomys   latidens            (Sanborn, 1952)                 NaN   
1      Abeillia   abeillei  (Lesson & DeLattre, 1839)                 NaN   
2   Abeomelomys      sevia    (Tate & Archbold, 1935)                 NaN   
3  Abrawayaomys    ruschii         Cunha & Cruz, 1979                 NaN   
4      Abrocoma  bennettii           Waterhouse, 1837                 NaN   

   ...                             Synonyms  \
0  ...                                  NaN   
1  ...                                  NaN   
2  ...  Melomys sevia|Pogonomelomys sevia     
3  ...                                  NaN   
4  ...                                  NaN   

                                  Common names (Eng)  Common names (Fre)  \
0                            Luzon Broad-toothed Rat                 NaN   
1                        Emerald-chinned Hummingbird                 NaN   
2  Menzies' Mouse, Menzies's Mouse, Highland Brus...                 NaN   
3                                       Ruschi's Rat                 NaN   
4                           Bennett's Chinchilla Rat                 NaN   

  Common names (Spa) Red List status Red List criteria  \
0                NaN              DD               NaN   
1                NaN              LC               NaN   
2                NaN              LC               NaN   
3                NaN              LC               NaN   
4                NaN              LC               NaN   

  Red List criteria version Year assessed Population trend  Petitioned  
0                       3.1        2008.0          unknown           N  
1                       3.1        2012.0       decreasing           N  
2                       3.1        2016.0          unknown           N  
3                       3.1        2016.0          unknown           N  
4                       3.1        2016.0          unknown           N  

[5 rows x 23 columns]

For both birds and mammals answer the following questions:

1. Size of the datasets.
The Flavivirus DataFrame has 570 rows and 10 columns.
The Birds Metadata DataFrame has 33265 rows and 17 columns.
The IUCN Mammals/Reptiles/Amphibians DataFrame has 16081 rows and 23 columns.

2. For birds, filter the dataset using species-level rows only. Hint. Column Subsp. Seq. value of 0 indicates the main species name.
DataFrame has 11204 rows and 17 columns.

3. How many bird species and mammal species are present in data frames?
There are 11198 unique bird species in the filtered bird dataset.
There are 8779 unique mammal species in the IUCN mammals dataset.

#Database Integration Exercise

In [ ]:
df_iucn = mammals_meta.copy()

print(f"'df' DataFrame has {df.shape[0]} rows and {df.shape[1]} columns.")
print(f"'birds_meta' DataFrame has {birds_meta.shape[0]} rows and {birds_meta.shape[1]} columns.")
print(f"'df_iucn' DataFrame has {df_iucn.shape[0]} rows and {df_iucn.shape[1]} columns.")

'df' DataFrame has 570 rows and 10 columns.
'birds_meta' DataFrame has 33265 rows and 17 columns.
'df_iucn' DataFrame has 16081 rows and 23 columns.


In [ ]:
birds_species_only = birds_meta[birds_meta['Subsp. Seq.'] == 0.0]

print("First 5 rows of the filtered birds_species_only DataFrame:")
print(birds_species_only.head())

print(f"\nShape of birds_species_only DataFrame: {birds_species_only.shape[0]} rows, {birds_species_only.shape[1]} columns")

First 5 rows of the filtered birds_species_only DataFrame:
    Seq.  Subsp. Seq.             Order    Family name     Family Subfamily  \
0    1.0          0.0  STRUTHIONIFORMES  Struthionidae  Ostriches       NaN   
5    2.0          0.0  STRUTHIONIFORMES  Struthionidae  Ostriches       NaN   
6    3.0          0.0  STRUTHIONIFORMES        Rheidae      Rheas       NaN   
12   4.0          0.0  STRUTHIONIFORMES        Rheidae      Rheas       NaN   
16   5.0          0.0  STRUTHIONIFORMES      Tinamidae   Tinamous       NaN   

   Tribe             Common name         Scientific name          Authority  \
0    NaN          Common Ostrich        Struthio camelus     Linnaeus, 1758   
5    NaN          Somali Ostrich  Struthio molybdophanes    Reichenow, 1883   
6    NaN            Greater Rhea          Rhea americana   (Linnaeus, 1758)   
12   NaN             Lesser Rhea            Rhea pennata    d'Orbigny, 1834   
16   NaN  Tawny-breasted Tinamou      Nothocercus julius  (Bonaparte, 1

In [ ]:
unique_bird_species_count = birds_species_only['Scientific name'].nunique()
print(f"Number of unique bird species: {unique_bird_species_count}")

unique_mammal_species_count = df_iucn['Species'].nunique()
print(f"Number of unique mammal species: {unique_mammal_species_count}")

Number of unique bird species: 11198
Number of unique mammal species: 8779


In [ ]:
birds_prepared = birds_species_only[['Scientific name', 'Order', 'Family name']].copy()
birds_prepared = birds_prepared.rename(columns={'Scientific name': 'Scientific Name', 'Family name': 'Family'})
birds_prepared['Class'] = 'Aves'

mammals_prepared = df_iucn[['Species', 'Order', 'Family', 'Class']].copy()
mammals_prepared = mammals_prepared.rename(columns={'Species': 'Scientific Name'})

print("Birds Prepared DataFrame (first 5 rows):")
print(birds_prepared.head())
print("\nBirds Prepared DataFrame Columns:")
print(birds_prepared.columns)

print("\nMammals Prepared DataFrame (first 5 rows):")
print(mammals_prepared.head())
print("\nMammals Prepared DataFrame Columns:")
print(mammals_prepared.columns)

Birds Prepared DataFrame (first 5 rows):
           Scientific Name             Order         Family Class
0         Struthio camelus  STRUTHIONIFORMES  Struthionidae  Aves
5   Struthio molybdophanes  STRUTHIONIFORMES  Struthionidae  Aves
6           Rhea americana  STRUTHIONIFORMES        Rheidae  Aves
12            Rhea pennata  STRUTHIONIFORMES        Rheidae  Aves
16      Nothocercus julius  STRUTHIONIFORMES      Tinamidae  Aves

Birds Prepared DataFrame Columns:
Index(['Scientific Name', 'Order', 'Family', 'Class'], dtype='object')

Mammals Prepared DataFrame (first 5 rows):
  Scientific Name             Order       Family     Class
0        latidens          RODENTIA      MURIDAE  MAMMALIA
1        abeillei  CAPRIMULGIFORMES  TROCHILIDAE      AVES
2           sevia          RODENTIA      MURIDAE  MAMMALIA
3         ruschii          RODENTIA   CRICETIDAE  MAMMALIA
4       bennettii          RODENTIA  ABROCOMIDAE  MAMMALIA

Mammals Prepared DataFrame Columns:
Index(['Scientific Nam

#Perform a vertical join of Mammals and Birds data frames:

In [ ]:
metadata = pd.concat([birds_prepared, mammals_prepared], axis=0, ignore_index=True)

print("First 5 rows of the combined metadata DataFrame:")
print(metadata.head())

print(f"\nShape of metadata DataFrame: {metadata.shape[0]} rows, {metadata.shape[1]} columns")

First 5 rows of the combined metadata DataFrame:
          Scientific Name             Order         Family Class
0        Struthio camelus  STRUTHIONIFORMES  Struthionidae  Aves
1  Struthio molybdophanes  STRUTHIONIFORMES  Struthionidae  Aves
2          Rhea americana  STRUTHIONIFORMES        Rheidae  Aves
3            Rhea pennata  STRUTHIONIFORMES        Rheidae  Aves
4      Nothocercus julius  STRUTHIONIFORMES      Tinamidae  Aves

Shape of metadata DataFrame: 27285 rows, 4 columns


#Merge the metadata dataframe with the Flavivirus dataframe:

In [ ]:
df_merged_inner = pd.merge(df, metadata, on='Scientific Name', how='inner')

print("First 5 rows of the merged DataFrame (df_merged_inner):")
print(df_merged_inner.head())

print(f"\nShape of df_merged_inner DataFrame: {df_merged_inner.shape[0]} rows, {df_merged_inner.shape[1]} columns")

First 5 rows of the merged DataFrame (df_merged_inner):
   sr no.                        Virus          Scientific Name  \
0       4                 Bagaza virus           Alectoris rufa   
1       5                 Bagaza virus      Phasianus colchicus   
2      55                 Ilheus virus    Molothrus bonariensis   
3      56                 Ilheus virus  Sporophila caerulescens   
4      79  Japanese encephalitis virus            Turdus merula   

                                             Authors  Year  \
0  D. Buitrago, A. Rocha, C. Tena-Tomas, M. Vigo,...  2012   
1  D. Buitrago, A. Rocha, C. Tena-Tomas, M. Vigo,...  2012   
2          L. Spence, C. R. Anderson and W. G. Downs  1962   
3          L. Spence, C. R. Anderson and W. G. Downs  1962   
4                                            GenBank   NaN   

                              Title/accession number  \
0  Real-time fluorogenic reverse transcription po...   
1  Real-time fluorogenic reverse transcription po...   


In [ ]:
df_merged_outer = pd.merge(df, metadata, on='Scientific Name', how='outer')

print("First 5 rows of the merged DataFrame (df_merged_outer):")
print(df_merged_outer.head())

print(f"\nShape of df_merged_outer DataFrame: {df_merged_outer.shape[0]} rows, {df_merged_outer.shape[1]} columns")

First 5 rows of the merged DataFrame (df_merged_outer):
   sr no. Virus           Scientific Name Authors Year Title/accession number  \
0     NaN   NaN         Abeillia abeillei     NaN  NaN                    NaN   
1     NaN   NaN    Abroscopus albogularis     NaN  NaN                    NaN   
2     NaN   NaN    Abroscopus schisticeps     NaN  NaN                    NaN   
3     NaN   NaN  Abroscopus superciliaris     NaN  NaN                    NaN   
4     NaN   NaN            Aburria aburri     NaN  NaN                    NaN   

  Journal  Issue Volume Pages             Order         Family Class  
0     NaN    NaN    NaN   NaN  CAPRIMULGIFORMES    Trochilidae  Aves  
1     NaN    NaN    NaN   NaN     PASSERIFORMES  Scotocercidae  Aves  
2     NaN    NaN    NaN   NaN     PASSERIFORMES  Scotocercidae  Aves  
3     NaN    NaN    NaN   NaN     PASSERIFORMES  Scotocercidae  Aves  
4     NaN    NaN    NaN   NaN       GALLIFORMES       Cracidae  Aves  

Shape of df_merged_outer DataF

In [ ]:
df_merged_left = pd.merge(df, metadata, on='Scientific Name', how='left')

print("First 5 rows of the merged DataFrame (df_merged_left):")
print(df_merged_left.head())

print(f"\nShape of df_merged_left DataFrame: {df_merged_left.shape[0]} rows, {df_merged_left.shape[1]} columns")

First 5 rows of the merged DataFrame (df_merged_left):
   sr no.         Virus     Scientific Name  \
0       3    Apoi virus       Rattus rattus   
1       2    Apoi virus    Myodes rufocanus   
2       0    Apoi virus  Apodemus argenteus   
3       1    Apoi virus  Apodemus speciosus   
4       4  Bagaza virus      Alectoris rufa   

                                             Authors  Year  \
0  C. M. Baak-Baak, N. Cigarroa-Toledo, O. M. Tor...  2016   
1                                        CDC Arbocat   NaN   
2                 J. S. Mackenzie and D. T. Williams  2009   
3                 J. S. Mackenzie and D. T. Williams  2009   
4  D. Buitrago, A. Rocha, C. Tena-Tomas, M. Vigo,...  2012   

                              Title/accession number  \
0  New Records of Ectoparasites Echinolaelaps ech...   
1                                        CDC Arbocat   
2  The zoonotic flaviviruses of southern, south-e...   
3  The zoonotic flaviviruses of southern, south-e...   
4  Real-t

In [ ]:
df_merged_right = pd.merge(df, metadata, on='Scientific Name', how='right')

print("First 5 rows of the merged DataFrame (df_merged_right):")
print(df_merged_right.head())

print(f"\nShape of df_merged_right DataFrame: {df_merged_right.shape[0]} rows, {df_merged_right.shape[1]} columns")

First 5 rows of the merged DataFrame (df_merged_right):
   sr no.            Virus         Scientific Name  \
0   522.0  West Nile virus        Struthio camelus   
1     NaN              NaN  Struthio molybdophanes   
2     NaN              NaN          Rhea americana   
3     NaN              NaN            Rhea pennata   
4     NaN              NaN      Nothocercus julius   

                      Authors  Year  \
0  M. Venter and R. Swanepoel  2010   
1                         NaN   NaN   
2                         NaN   NaN   
3                         NaN   NaN   
4                         NaN   NaN   

                              Title/accession number  \
0  West Nile virus lineage 2 as a cause of zoonot...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                     Journal  Issue Volume   Pa

In [ ]:
df_merged = df_merged_left.copy()

print("Reasoning for choosing Left Join:\n")
print("The primary goal is to combine the Flavivirus data (df) with relevant host metadata (metadata) for further analysis, specifically enriching the Flavivirus host data with taxonomic information while ensuring that ALL original Flavivirus records are retained. A left join accomplishes this by keeping all rows from the left DataFrame (df) and adding matching information from the right DataFrame (metadata). Where no match is found in the metadata for a given scientific name in df, the taxonomic columns (Order, Family, Class) are filled with NaN, indicating missing metadata rather than discarding the Flavivirus record itself.\n")

print("First 5 rows of the final merged DataFrame (df_merged):\n")
print(df_merged.head())

print(f"\nShape of df_merged DataFrame: {df_merged.shape[0]} rows, {df_merged.shape[1]} columns")

Reasoning for choosing Left Join:

The primary goal is to combine the Flavivirus data (df) with relevant host metadata (metadata) for further analysis, specifically enriching the Flavivirus host data with taxonomic information while ensuring that ALL original Flavivirus records are retained. A left join accomplishes this by keeping all rows from the left DataFrame (df) and adding matching information from the right DataFrame (metadata). Where no match is found in the metadata for a given scientific name in df, the taxonomic columns (Order, Family, Class) are filled with NaN, indicating missing metadata rather than discarding the Flavivirus record itself.

First 5 rows of the final merged DataFrame (df_merged):

   sr no.         Virus     Scientific Name  \
0       3    Apoi virus       Rattus rattus   
1       2    Apoi virus    Myodes rufocanus   
2       0    Apoi virus  Apodemus argenteus   
3       1    Apoi virus  Apodemus speciosus   
4       4  Bagaza virus      Alectoris rufa 

#Pivot table

In [ ]:
df_merged['Order'] = df_merged['Order'].fillna('Unknown Order')

pivot_table_host_diversity = pd.pivot_table(df_merged,
                                            index='Virus',
                                            columns='Order',
                                            values='Scientific Name',
                                            aggfunc='nunique')

pd.set_option('display.max_columns', None)
print("Pivot Table of Host Species Diversity by Virus and Order:\n")
print(pivot_table_host_diversity)
pd.reset_option('display.max_columns')

Pivot Table of Host Species Diversity by Virus and Order:

Order                             ACCIPITRIFORMES  ANSERIFORMES  \
Virus                                                             
Apoi virus                                    NaN           NaN   
Bagaza virus                                  NaN           NaN   
Banzi virus                                   NaN           NaN   
Bouboui virus                                 NaN           NaN   
Bussuquara virus                              NaN           NaN   
Cacipacore virus                              NaN           NaN   
Dakar bat virus                               NaN           NaN   
Dengue virus                                  NaN           NaN   
Entebbe bat virus                             NaN           NaN   
Iguape virus                                  NaN           NaN   
Ilheus virus                                  NaN           NaN   
Japanese encephalitis virus                   NaN           NaN   
Kou

#concept map

Step-by-Step Instructions to Create the Concept Map:

Choose a Diagramming Tool: Open your preferred diagramming tool (e.g., Google Slides, PowerPoint, LucidChart, Draw.io, or even a pen and paper).

Represent Each DataFrame as a Box:

Draw three distinct boxes, one for each of your main DataFrames:

df (Flavivirus data)
birds_meta (Bird metadata)
df_iucn (Mammal/Reptile/Amphibian metadata)
Label each box clearly with the DataFrame's name. For example:

+-----------------+
|       df        |
| (Flavivirus Data)|
+-----------------+
List Key Columns and Identify Primary Keys within Each Box:

Inside each DataFrame's box, list its relevant columns. Pay special attention to columns that contain unique identifiers or names that will be used for merging.

For df (Flavivirus Data):

Relevant column: Scientific Name (This will be a key for joining with metadata).
Identify sr no. as a potential primary key.
For birds_meta (Bird Metadata):

Relevant columns: Scientific name, Order, Family name, Class (which we added as 'Aves').
Identify SISRecID or SpcRecID as potential primary keys for unique species identification, and 'Scientific name' as a join key.
For df_iucn (Mammal/Reptile/Amphibian Metadata):

Relevant columns: Species, Order, Family, Class.
Identify Species ID as a potential primary key, and 'Species' as a join key.
Clearly mark or highlight the Primary Keys (e.g., by underlining them or writing 'PK' next to them).

Illustrate the Vertical Join (Concatenation) of birds_meta and df_iucn:

Draw an arrow or connector from the birds_meta box and another from the df_iucn box, both leading to a new box representing the metadata DataFrame.

Indicate the type of operation (e.g., 'Vertical Join' or 'pd.concat()').

In the metadata box, list the common columns used in the concatenation: Scientific Name, Order, Family, Class.

+-----------------+       +-------------------+
|   birds_meta    |       |     df_iucn       |
| (Bird Metadata) |       | (Mammal Metadata) |
+-----------------+       +-------------------+
        |                          |
        |        (Vertical Join / pd.concat())       |
        v                          v
+-----------------------------------+
|              metadata             |
| (Scientific Name, Order, Family, Class) |
+-----------------------------------+
Illustrate the Horizontal Join (Merge) of df and metadata:

Draw an arrow or connector from the df box and another from the metadata box, both leading to a final box representing the df_merged DataFrame.

Indicate the join key used: Scientific Name.

Specify the type of join you chose as most appropriate (Left Join).

+-----------------+
|       df        |
| (Flavivirus Data)|
+-----------------+
        |           (Left Join on 'Scientific Name')
        |                          |
        v                          v
+-------------------------------------------------------+
|                      df_merged                        |
| (All columns from df + Order, Family, Class from metadata) |
+-------------------------------------------------------+
Add Notes and Explanations:

Include small text boxes or labels next to arrows to explain the type of join (inner, outer, left, right) explored and why the 'Left Join' was chosen for df_merged.
Mention the implications of each join type, especially the data loss in inner join or NaNs in left/outer joins.
Review and Refine:

Ensure all relationships are clear and easy to understand.
Check that all primary keys and join keys are correctly identified.
Make sure the flow of data is logically represented.
This will provide a clear visual understanding of how the dataframes are connected and how the integration was performed.